# Тести для малих вибірок та пропорцій



**Завдання 1**. E-commerce компанія після редизайну сайту підозрює, що **середній час до покупки (time-to-purchase)** користувачів **збільшився**.
Історично середній час від першого заходу на сайт до покупки становив $\mu_0 = 123.7$ хвилин.

Команда аналітиків випадково обрала дані **7 покупців після редизайну**:
`128, 135, 121, 142, 126, 151, 123`

З допомогою наявних даних зʼясуйте, чи збільшився середній час покупки після редизайну.

Для цього виконайте наступні 6 кроків. Правильне виконання кожного з кроків оцінюється в 1 бал.

1. Запишіть нульову та альтернативну гіпотези і визначте тип тесту.
2. Обчисліть вибіркові статистики: $\bar x$, $s$, $n$.
3. Оберіть тип тесту та виконайте його (знайдіть критичне значення тесту, статистику тесту та р-значення) будь-яким пасуючим способом, щоб перевірити гіпотезу на рівні значущості $\alpha = 0.10$.
4. Прийміть рішення, чи відхиляєте ви гіпотезу $H_0$ використовуючи p-value.
5. Напишіть висновок: чи справді редизайн сайту подовжив час до покупки?
6. Чи зміниться ваше рішення при зміні рівня значущості на $\alpha = 0.05$.


In [1]:
import numpy as np
from scipy import stats
import statsmodels.api as sm
from statsmodels.stats.proportion import proportions_ztest, proportion_confint

**Розв'язок:**

1. Визначаю нульову та альтернативну гіпотези та тип тесту:

    - Нульова гіпотеза $H_о$ : середній час від першого заходу на сайт до покупки становить 123.7 хвилин - $\mu_о$ = 123.7.
    - Альтернативна гіпотеза $H_a$ : середній час до покупки збільшився - $\mu_a$ > 123.7.
    - Це правосторонній тест.

2. Обчислюю вибіркові статистики: $\bar x$, $s$, $n$:

In [19]:
# Дані з умови (вибірка після редизайну)
time = np.array([128, 135, 121, 142, 126, 151, 123])
mu0 = 123.7   # середній час до покупки
alpha = 0.10

# Обчислюю вибіркову статистику
x_bar = np.mean(time)
s = np.std(time, ddof=1)
n = len(time)

# Виводжу результати
print(f"Вибіркове середнє: {x_bar:.2f}")
print(f"Вибіркове стандартне відхилення: {s:.2f}")
print(f"Розмір вибірки: {n}")

Вибіркове середнє: 132.29
Вибіркове стандартне відхилення: 10.98
Розмір вибірки: 7


3. Шукаю статистику тесту, критичне значення тесту та р-value:

In [15]:
# Обчислюю t-статистику
t_stat = (x_bar - mu0) / (s / np.sqrt(n))

# Критичне значення для правостороннього тесту
df = n - 1
t_crit = stats.t.ppf(1 - alpha, df)

# Обчислюю p-value
p_value = 1 - stats.t.cdf(t_stat, df)


# Виводжу результати
print(f"t-статистика: {t_stat:.3f}")
print(f"Критичне значення (df={df}): {t_crit:.3f}")
print(f"Порівняння: {t_stat:.3f} {'<' if t_stat < t_crit else '≥'} {t_crit:.3f}")

# Висновок
if t_stat > t_crit:
    print("Висновок: t-статистика ПЕРЕВИЩУЄ критичне значення.")
    decision1 = "ВІДХИЛЯЄМО H₀: середній час до покупки збільшився на рівні значущості 0.1."
else:
    print("Висновок: t-статистика НЕ перевищує критичне значення.")
    decision1 = "НЕ ВІДХИЛЯЄМО H₀."

print(f"Рішення: {decision1}")

t-статистика: 2.069
Критичне значення (df=6): 1.440
Порівняння: 2.069 ≥ 1.440
Висновок: t-статистика ПЕРЕВИЩУЄ критичне значення.
Рішення: ВІДХИЛЯЄМО H₀: середній час до покупки збільшився на рівні значущості 0.1.


4. Роблю висновок щодо відхилення нульової гіпотези $H_0$ використовуючи p-value:

In [13]:
print(f"p-value = {p_value:.4f}")
print(f"Рівень значущості α = {alpha}")
print(f"Порівняння: {p_value:.4f} {'<' if p_value < alpha else '≥'} {alpha}")

if p_value < alpha:
    print("Висновок: p-value МЕНШЕ за рівень значущості.")
    decision2 = "ВІДХИЛЯЄМО H₀: середній час до покупки збільшився на рівні значущості 0.1."
else:
    print("Висновок: p-value НЕ менше за рівень значущості.")
    decision2 = "НЕ ВІДХИЛЯЄМО H₀."

print(f"Рішення: {decision2}")

p-value = 0.0420
Рівень значущості α = 0.1
Порівняння: 0.0420 < 0.1
Висновок: p-value МЕНШЕ за рівень значущості.
Рішення: ВІДХИЛЯЄМО H₀: середній час до покупки збільшився на рівні значущості 0.1.


  **5. Загальний висновок:**

Завдяки двом незалежним способам перевірки, через прорахунок t‑статистики та p‑value, підтвердилось одне й те саме:

- t‑статистика = 2.069 означає, що вибіркове середнє більше за історичне, і відхилення більше за критичне значення t_crit=1.440.
- p‑value = 0.0420 НЕ перевищує рівень значущості $\alpha = 0.10$, тобто то ймовірність отримати такі дані при $H_о$ дуже мала.

**Відхиляємо нульову гіпотезу $H_о$**. Це означає, що середній час до покупки статистично значущо збільшився на рівні значущості 0.1.

6. Роблю перевірку стосовно зміни рішення і висновків при зміні рівня значущості на $\alpha = 0.05$ на основі порівняння з **p-value**.

In [17]:
alpha_alt = 0.05

print(f"p-value = {p_value:.4f}")
print(f"Рівень значущості α = {alpha_alt}")
print(f"Порівняння: {p_value:.4f} {'<' if p_value < alpha_alt else '≥'} {alpha_alt}")

if p_value < alpha_alt:
    print("Висновок: p-value МЕНШЕ за рівень значущості.")
    decision2 = "ВІДХИЛЯЄМО H₀: середній час до покупки збільшився на рівні значущості 0.05."
else:
    print("Висновок: p-value НЕ менше за рівень значущості.")
    decision2 = "НЕ ВІДХИЛЯЄМО H₀."

print(f"Рішення: {decision2}")

p-value = 0.0420
Рівень значущості α = 0.05
Порівняння: 0.0420 < 0.05
Висновок: p-value МЕНШЕ за рівень значущості.
Рішення: ВІДХИЛЯЄМО H₀: середній час до покупки збільшився на рівні значущості 0.05.


**Висновок:**

p‑value = 0.042 все одно НЕ перевищує новий рівень значущості $\alpha = 0.05$, тобто на рівні значущості 5% є статистично значущі докази того, що після редизайну сайту середній час до покупки збільшився.
Тож **відхиляємо нульову гіпотезу $H_о$**. 

Але сила доказів залежить саме від обраного рівня значущості. При більш серйозній зміні, наприклад, на $\alpha = 0.01$, доказів було б недостатньо і нульова гіпотеза не відхилялася.

**Завдання 2.**

До спеціальної рекламної кампанії **23%** дорослих упізнавали логотип компанії. Після завершення кампанії відділ маркетингу провів опитування: з **1200** випадково відібраних дорослих **311** упізнали логотип.

Перевірте на рівні значущості **3%** ($\alpha=0.03$), чи дають ці дані достатні підстави стверджувати, що **тепер більше ніж 23%** дорослих упізнають логотип компанії. Для розвʼязку використовуйте бібліотеку `statsmodels`.

Зробіть висновок, чи зросла впізнаваність логотипу.

Додатково, обчисліть довірчий інтревал на заданому рівні значущості і проінтерпретуйте текстом - як він додатково пояснює прийняте нами рішення?

**Розв'язок:**

Визначаю нульову та альтернативну гіпотези та тип тесту:

- Нульова гіпотеза $H_о$ : частка дорослих, які впізнають логотип, залишилась на рівні 23% - $p_o$ = 0,23.
- Альтернативна гіпотеза $H_a$ : середній час до покупки збільшився - $p_a$ > 0,23.
- Це правосторонній тест пропорцій.

In [21]:
count = 311       # кількість "успіхів"
nobs = 1200       # обсяг вибірки
p0 = 0.23         # гіпотетична пропорція (Ho)
alpha2 = 0.03     # рівень значущості

# z-тест для пропорції
stat, pval = proportions_ztest(count, nobs, p0, alternative='larger')
confint = proportion_confint(count, nobs, alpha2, method='normal')

print("Тест для однієї пропорції:")
print(f"Z-статистика = {stat:.3f}")
print(f"p-value = {pval:.4f}")
print(f"95% довірчий інтервал: {confint}")
print(f"Рівень значущості α = {alpha2}")
print(f"Порівняння: {pval:.4f} {'<' if pval < alpha2 else '≥'} {alpha2}")

if pval < alpha2:
    print("Висновок: p-value МЕНШЕ за рівень значущості.")
    decision3 = "ВІДХИЛЯЄМО H₀: є статистично значущі докази того, що впізнаваність логотипу зросла понад 23%."
else:
    print("Висновок: p-value НЕ менше за рівень значущості.")
    decision3 = "НЕ ВІДХИЛЯЄМО H₀."

print(f"Рішення: {decision3}")

Тест для однієї пропорції:
Z-статистика = 2.306
p-value = 0.0106
95% довірчий інтервал: (0.23171700302179205, 0.28661633031154127)
Рівень значущості α = 0.03
Порівняння: 0.0106 < 0.03
Висновок: p-value МЕНШЕ за рівень значущості.
Рішення: ВІДХИЛЯЄМО H₀: є статистично значущі докази того, що впізнаваність логотипу зросла понад 23%.


**Висновок по довірчому інтервалу:**
95% довірчий інтервал для пропорції впізнаваності логотипу знаходиться в межах (0.232, 0.287).
Оскільки весь інтервал лежить вище порогового значення 0.23, можна стверджувати, що з імовірністю 95% справжній рівень впізнаваності вище 23%.